sdk/containerregistry/azure-containers-containerregistry/src/main/java/com/azure/containers/containerregistry/implementation/authentication/ContainerRegistryCredentialsPolicy.java

```diff
@@ -0,0 +1,154 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.containers.containerregistry.implementation.authentication;
+
+import com.azure.core.credential.TokenCredential;
+import com.azure.core.http.HttpPipeline;
+import com.azure.core.http.HttpPipelineCallContext;
+import com.azure.core.http.HttpPipelineNextPolicy;
+import com.azure.core.http.HttpResponse;
+import com.azure.core.http.policy.HttpPipelinePolicy;
+import com.azure.core.util.logging.ClientLogger;
+import com.azure.core.util.serializer.SerializerAdapter;
+import reactor.core.publisher.Mono;
+
+
+import java.util.Map;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.HashMap;
+
+/**
+ * <p>Credential policy for the container registry. It follows the challenge based authorization scheme.</p>
+ *
+ * <p>For example GET /api/v1/acr/repositories translates into the following calls.</p>
+ *
+ * <p>Step1: GET /api/v1/acr/repositories
+ * Return Header: 401: www-authenticate header - Bearer realm=&quot;{url}&quot;,service=&quot;{serviceName}&quot;,scope=&quot;{scope}&quot;,error=&quot;invalid_token&quot;.</p>
+ *
+ * <p>Step2: Parse the serviceName, scope from the service.</p>
+ *
+ * <p>Step3: POST /api/oauth2/exchange  Request Body : {service, scope, grant-type, aadToken with ARM scope}
+ * Response Body: {acrRefreshToken}</p>
+ *
+ * <p>Step4: POST /api/oauth2/token  Request Body: {acrRefreshToken, scope, grant-type}
+ * Response Body: {acrAccessToken}</p>
+ *
+ * <p>Step5: GET /api/v1/acr/repositories
+ * Request Header: {Bearer acrTokenAccess}</p>
+ */
+public class ContainerRegistryCredentialsPolicy implements HttpPipelinePolicy {
```

Comment:
Generally, we make all pipeline policies public so that users add the policy to their own HTTP pipeline.

Reply:
The policy is not public - just because we are still ironing out what is the right handshake with experimental package. Having said that, this policy may not be public even in future - since its a convenience layer added for the customer. 

---
In reply to: [597957262](https://github.com/Azure/azure-sdk-for-java/pull/19926#discussion_r597957262) [](ancestors = 597957262)